Deprecate TypedStorage, its derived classes, and all of their public methods

kurtamohler opened the pull request:
Part of #85302


cc @albanD @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire @jansel @lezcano @fdrocha

## BC-breaking note

### Deprecate `torch.Tensor.storage()` in favor of `torch.Tensor.untyped_storage()`

Version 1.13

```python
tensor.storage()
```

Version 2.0

```python
tensor.untyped_storage()
```

### Deprecate `torch.TypedStorage` and all its methods in favor of `torch.UntypedStorage`

Version 1.13

```python
torch.TypedStorage(...)
```

Version 2.0

```python
torch.UntypedStorage(...)
```

If you need to access individual elements in a storage as a particular dtype, you can simply create a tensor to view it:

```python
torch.tensor(storage, dtype=...)
```

kurtamohler committed:
Deprecate TypedStorage, its derived classes, and all of their public methods

ezyang reviewed (approved):
Can you doublecheck that the deprecation warning only gets raised once? Thanks!

kurtamohler commented:
@ezyang, I've added a test that it only gets raised once unless warnings are cleared. I didn't add all of the functions to the test, but I can if we want to be that thorough

ezyang commented:
nah that's good enough

ezyang commented:
A good follow up would be to make sure pytorch proper doesn't raise these deprecations

ezyang commented:
@pytorchbot merge

kurtamohler commented:
> A good follow up would be to make sure pytorch proper doesn't raise these deprecations

By this, do you mean because a `DeprecationWarning` doesn't raise by default? Actually, I forgot to use DeprecationWarning, I will fix that

ezyang commented:
> By this, do you mean because a DeprecationWarning doesn't raise by default? Actually, I forgot to use DeprecationWarning, I will fix that

Different: what I mean is that we shouldn't trigger the deprecation warning if the user didn't explicitly use typed storage. If pytorch internally is hitting the DeprecationWarning, we should fix it (because otherwise it will spam users)

kurtamohler commented:
I see, good point

kurtamohler commented:
@ezyang , there are some places where internal calls are raising the warning. For instance, when serializing tensors.

I'm not sure what would be the best way to avoid internally generated warnings, but I thought of two options.

The first one is to use the `warnings` module filter to suppress the warning at every internal call site. But I don't think that's very good, since the filter has to do a string search and it will probably affect performance significantly.

The other is to add an underscored version of each function which does everything except for raising the warning. The public function would just raise the warning and then call the underscored function. We would change all the internal call sites to use the underscored version to avoid raising the warning. Something like this:

```python
def _func(...):
    # do stuff

def func(...):
    _warn_typed_storage_removal()
    return _func(...)
```

I think I like this solution, but what do you think? Is there a better way? I tried googling a common solution for this and haven't found anything yet

ezyang commented:
No warn variants sgtm, esp if you only need a few / we have a strategy for getting rid of them

kurtamohler commented:
Actually, I just realized that I can't do this for dunder functions. So instead, I think I'll have to add a kwarg `_internal=False` to all the functions. If it's False, the warning will get raised

kurtamohler committed:
Avoid raising warnings for internal TypedStorage usage

kurtamohler commented:
I've fixed a lot of the places where internal storage method calls were raising the warning. There may be more though, because I don't really have a great way to find all of them. I found some call sites with grep and some by running a bunch of the unit tests that don't deal with storages directly and raise an error if `_warn_typed_storage_removal` is called.

Do you think this is alright, @ezyang , or should I keep looking for more internal call sites?

(I also renamed a couple things, like `TypedStorage._storage` --> `TypedStorage._untyped_storage` to make them a bit clearer)

ezyang commented:
I'm happy to ship this as is. Do you want another look over?

ezyang commented:
looks like you've still got some tests to fix

kurtamohler commented:
Well, I think I need to suppress more of the warnings before we can merge this. In the [linux-bionic-py3.7-clang9 / test (crossref, 2, 2, linux.2xlarge)](https://github.com/pytorch/pytorch/actions/runs/3191200993/jobs/5207408145#logs) job, 48% of the lines in the log are the warning message, which is way too many

ezyang commented:
> One thing that would reduce the number of warnings is to use the internal TypedStorage functions in the unit tests instead of the public ones. I'm not sure if that's a good idea though. Or I could add a warning filter to tests that use the public functions

For the tests, it's fine to suppress them entirely, since you're testing deprecated apis, the deprecation doesn't apply.

kurtamohler commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1999,6 +1999,7 @@ def wrap_with_cuda_memory_check(self, method):
     # show up red).
     # Otherwise, the flaky test will show up green while its stats are captured by test reports.
     def _run_with_retry(self, result=None, num_runs_left=0, report_only=True, num_red=0, num_green=0):
+        warnings.filterwarnings('ignore', 'TypedStorage is deprecated')
```
@ezyang, do you think it's alright to do this? It does suppress the warnings from tests, at least locally for me. I'll double check the CI logs after they finish

ezyang replied:
ok. But there's also something else I'd like you to investigate, which is why the warnings keep popping up repeatedly; they should only trigger once per process and then not again.

kurtamohler replied:
So it looks like `test_proxy_tensor.py`, with `PYTORCH_TEST_WITH_CROSSREF=1` has a few places where different warnings (not just the TypedStorage deprecation warning) are getting duplicated. Here's one example:

<details><summary>click to expand</summary>

```
$ PYTORCH_TEST_WITH_CROSSREF=1 python test/test_proxy_tensor.py -k test_make_fx_fake_exhaustive_nn_functional_upsample_nearest_cpu_float32
/work2/kurtamohler/development/pytorch-0/torch/backends/cudnn/__init__.py:91: UserWarning: PyTorch was compiled without cuDNN/MIOpen support. To use cuDNN/MIOpen, rebuild PyTorch making sure the library is visible to the build system.
  warnings.warn(
/work2/kurtamohler/development/pytorch-0/test/test_proxy_tensor.py:124: UserWarning: Couldn't import torchvision. Some of our tests use it, try to install it with commands from pytorch.org, post-fixed with `--no-deps` to avoid overwriting the pytorch installation
  warnings.warn("Couldn't import torchvision. Some of our tests use it, try "
/work2/kurtamohler/development/pytorch-0/torch/nn/functional.py:4014: UserWarning: nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.
  warnings.warn("nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.")
/work2/kurtamohler/development/pytorch-0/torch/nn/functional.py:4014: UserWarning: nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.
  warnings.warn("nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.")
/work2/kurtamohler/development/pytorch-0/torch/nn/functional.py:4014: UserWarning: nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.
  warnings.warn("nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.")
/work2/kurtamohler/development/pytorch-0/torch/nn/functional.py:4014: UserWarning: nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.
  warnings.warn("nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.")
/work2/kurtamohler/development/pytorch-0/torch/nn/functional.py:4014: UserWarning: nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.
  warnings.warn("nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.")
/work2/kurtamohler/development/pytorch-0/torch/nn/functional.py:4014: UserWarning: nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.
  warnings.warn("nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.")
/work2/kurtamohler/development/pytorch-0/torch/nn/functional.py:4014: UserWarning: nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.
  warnings.warn("nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.")
/work2/kurtamohler/development/pytorch-0/torch/nn/functional.py:4014: UserWarning: nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.
  warnings.warn("nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.")
/work2/kurtamohler/development/pytorch-0/torch/nn/functional.py:4014: UserWarning: nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.
  warnings.warn("nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.")
/work2/kurtamohler/development/pytorch-0/torch/nn/functional.py:4014: UserWarning: nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.
  warnings.warn("nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.")
/work2/kurtamohler/development/pytorch-0/torch/nn/functional.py:4014: UserWarning: nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.
  warnings.warn("nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.")
/work2/kurtamohler/development/pytorch-0/torch/nn/functional.py:4014: UserWarning: nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.
  warnings.warn("nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.")
/work2/kurtamohler/development/pytorch-0/torch/nn/functional.py:4014: UserWarning: nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.
  warnings.warn("nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.")
/work2/kurtamohler/development/pytorch-0/torch/nn/functional.py:4014: UserWarning: nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.
  warnings.warn("nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.")
/work2/kurtamohler/development/pytorch-0/torch/nn/functional.py:4014: UserWarning: nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.
  warnings.warn("nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.")
/work2/kurtamohler/development/pytorch-0/torch/nn/functional.py:4014: UserWarning: nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.
  warnings.warn("nn.functional.upsample_nearest is deprecated. Use nn.functional.interpolate instead.")
.
----------------------------------------------------------------------
Ran 1 test in 0.080s

OK
```

</details>

I'm not sure why this happens yet, but I was able to fix the places where the TypedStorage warning was being emitted (at least locally--we'll see what happens in CI) by correcting the `handle_torch_function` call in the definition of `Tensor._storage` in  `tools/autograd/templates/python_variable_methods.cpp`. It was `return handle_torch_function(self, "storage");`, but should be `return handle_torch_function(self, "_storage");` instead.

I think it probably makes sense to look into the crossref warning duplication in another issue

kurtamohler commented:
For some reason I'm not able to ssh into CI jobs anymore. Looking into it

kurtamohler commented:
I can't ssh into CI jobs because Quansight's `pytorch-jumphost.quansight.dev` server is down. While that's being fixed, I decided to put a short stack trace into the warning messages so I can get full information about where the duplicate warnings are coming from just from the logs.

kurtamohler commented:
There's are two errors that keeps coming up in the [linux-bionic-py3.7-clang9 / test (crossref, 1, 2, linux.2xlarge)](https://github.com/pytorch/pytorch/actions/runs/3284786311/jobs/5411409048#logs) CI job, which appear to be part of the command `/opt/conda/bin/python -bb test_serialization.py -v --import-slow-tests --import-disabled-tests`:

<details>
<summary>click to expand</summary>

```
======================================================================
FAIL [0.004s]: test_serialization_save_warnings (__main__.TestOldSerialization)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_serialization.py", line 417, in test_serialization_save_warnings
    self.assertEqual(len(warns), 0)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 2510, in assertEqual
    msg=(lambda generated_msg: f"{generated_msg} : {msg}") if isinstance(msg, str) and self.longMessage else msg,
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_comparison.py", line 1118, in assert_equal
    raise error_metas[0].to_error(msg)
AssertionError: Scalars are not equal!
Absolute difference: 1
Relative difference: inf
======================================================================
FAIL [0.002s]: test_serialization_save_warnings (__main__.TestSerialization)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_serialization.py", line 417, in test_serialization_save_warnings
    self.assertEqual(len(warns), 0)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 2510, in assertEqual
    msg=(lambda generated_msg: f"{generated_msg} : {msg}") if isinstance(msg, str) and self.longMessage else msg,
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_comparison.py", line 1118, in assert_equal
    raise error_metas[0].to_error(msg)
AssertionError: Scalars are not equal!
Absolute difference: 1
Relative difference: inf
```

</details>

But for some reason when I log into that CI machine and run `/opt/conda/bin/python -bb test_serialization.py -v --import-slow-tests --import-disabled-tests`, I don't get these errors. Maybe there's some environment state that the CI job sets up before running the tests that I'm not aware of.

kurtamohler commented:
One of the CI jobs is failing because [MultiPy depends on `TypedStorage._storage`](https://github.com/pytorch/multipy/blob/e4e279acd73a608e16379e2e7917489717b025ba/multipy/utils/_deploy.py#L40), which I renamed to `TypedStorage._untyped_storage`. Maybe it's simplest if I just restore the old name. Although the alternative is to just change MultiPy to use the public method `TypedStorage.untyped()` instead

kurtamohler commented:
My last update fixed most of the duplicate TypedStorage warnings, at least for the `linux-bionic-py3.7-clang9 / test (crossref, 1, 2, linux.2xlarge)` CI job. There are a handful left that I need to look into, and I'll also look at all the other CI logs as well

kurtamohler commented:
Some of the duplicate warnings happen because `gradcheck` evidently resets the warnings. For instance:

```python
import torch
import warnings
from torch.testing._internal.common_utils import gradcheck

for _ in range(3):
    torch.FloatStorage()
    warnings.warn('other warning')
    gradcheck(lambda x: x, [torch.zeros(10, dtype=torch.double, requires_grad=True)])
```

Output:
```
/work2/kurtamohler/development/pytorch-0/torch/storage.py:322: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.
  warnings.warn(message, UserWarning)
/work2/kurtamohler/development/pytorch-0/../tmp/tmp.py:7: UserWarning: other warning
  warnings.warn('other warning')
/work2/kurtamohler/development/pytorch-0/torch/storage.py:322: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.
  warnings.warn(message, UserWarning)
/work2/kurtamohler/development/pytorch-0/../tmp/tmp.py:7: UserWarning: other warning
  warnings.warn('other warning')
/work2/kurtamohler/development/pytorch-0/torch/storage.py:322: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.
  warnings.warn(message, UserWarning)
/work2/kurtamohler/development/pytorch-0/../tmp/tmp.py:7: UserWarning: other warning
  warnings.warn('other warning')

```

I'm not sure yet where in `gradcheck` warnings are getting reset. But this only happens in a handful of places in the tests, so I think it can looked into later.

Also, I'm still fixing up a few other internally generated warnings. Very close to being finished, I think

ezyang commented:
Weird, gradcheck shouldnt reset warnings. @albanD @soulitzer do either of you know?

kurtamohler commented:
Turns out `gradcheck` has this behavior just because of the `warnings.catch_warnings()` context manager: https://github.com/pytorch/pytorch/blob/7f88934a8fb9b376b32c722ac2f05959da34c147/torch/autograd/gradcheck.py#L835

For instance:
```python
import torch
import warnings

for _ in range(3):
    with warnings.catch_warnings() as w:
        torch.FloatStorage()
        warnings.warn('other warning')
```

Output:
```
/work2/kurtamohler/development/pytorch-0/torch/storage.py:321: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.
  warnings.warn(message, UserWarning)
/work2/kurtamohler/development/pytorch-0/../tmp1.py:7: UserWarning: other warning
  warnings.warn('other warning')
/work2/kurtamohler/development/pytorch-0/torch/storage.py:321: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.
  warnings.warn(message, UserWarning)
/work2/kurtamohler/development/pytorch-0/../tmp1.py:7: UserWarning: other warning
  warnings.warn('other warning')
/work2/kurtamohler/development/pytorch-0/torch/storage.py:321: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.
  warnings.warn(message, UserWarning)
/work2/kurtamohler/development/pytorch-0/../tmp1.py:7: UserWarning: other warning
  warnings.warn('other warning')
```

kurtamohler commented:
Looking through the logs, the warnings aren't very noisy anymore. Most of the test logs have less than 30 of them, and most of these are just from separate test file executions (separate processes). Only two of the jobs have more than 100 warnings.

I am sifting through all the logs now and soon I will post an overview of all the duplicate counts for each job and what I know about them. I think there's only a handful of root causes left now, and I think almost all of them can probably be investigated/fixed after this PR is merged (for instance, the duplicates from in-between `warnings.catch_warnings()` calls).

But I am sifting through the logs that generate the most duplicates now, and I'll post a summary soon

ezyang commented:
I think this is good enough, thank you for the detailed investigation. The `catch_warnings` footgun is I bet the biggest culprit, I wonder if there's a way we can workaround it in userland.

kurtamohler commented:
I might as well post what I have right now.


Below are counts for how many times the warning was raised for each of the test CI jobs in the inductor, periodic, pull, and trunk workflows.


<details><summary>inductor workflow</summary>

| warnings raised | log file name                                                      |
| --------------- | ------------------------------------------------------------------ |
| 0               | `1_cuda11.6-py3.10-gcc7-sm86  test (inductor, 1, 7, linux.g5..txt` |
| 0               | `2_cuda11.6-py3.10-gcc7-sm86  test (inductor, 2, 7, linux.g5..txt` |
| 0               | `3_cuda11.6-py3.10-gcc7-sm86  test (inductor, 3, 7, linux.g5..txt` |
| 0               | `4_cuda11.6-py3.10-gcc7-sm86  test (inductor, 4, 7, linux.g5..txt` |
| 0               | `5_cuda11.6-py3.10-gcc7-sm86  test (inductor, 5, 7, linux.g5..txt` |
| 0               | `5_test.txt`                                                       |
| 0               | `6_cuda11.6-py3.10-gcc7-sm86  test (inductor, 6, 7, linux.g5..txt` |
| 0               | `7_cuda11.6-py3.10-gcc7-sm86  test (inductor, 7, 7, linux.g5..txt` |

</details>

<details><summary>periodic workflow</summary>

| warnings raised | log file name                                                      |
| --------------- | ------------------------------------------------------------------ |
| 0               | `1_buck-build-test  buck-build-test.txt`                           |
| 0               | `1_linux-bionic-cuda11.6-py3.7-gcc7-debug  test (default, 1,.txt`  |
| 0               | `1_linux-bionic-cuda11.7-py3.7-gcc7-debug  test (default, 1,.txt`  |
| 0               | `1_linux-focal-rocm5.2-py3.8-slow  test (slow, 1, 1, linux.ro.txt` |
| 14              | `2_win-vs2019-cuda11.7-py3  test (default, 2, 2, windows.8xla.txt` |
| 15              | `3_linux-bionic-cuda11.7-py3.7-gcc7-debug  test (default, 3,.txt`  |
| 18              | `4_linux-bionic-cuda11.6-py3.7-gcc7-debug  test (default, 4,.txt`  |
| 19              | `3_linux-bionic-cuda11.6-py3.7-gcc7-debug  test (default, 3,.txt`  |
| 20              | `1_linux-focal-rocm5.2-py3.8-distributed  test (distributed,.txt`  |
| 22              | `2_linux-bionic-cuda11.7-py3.7-gcc7-debug  test (default, 2,.txt`  |
| 27              | `3_win-vs2019-cuda11.7-py3  test (force_on_cpu, 1, 1, windows.txt` |
| 30              | `2_linux-bionic-cuda11.6-py3.7-gcc7-debug  test (default, 2,.txt`  |
| 30              | `4_linux-bionic-cuda11.7-py3.7-gcc7-debug  test (default, 4,.txt`  |
| 36              | `2_linux-focal-rocm5.2-py3.8-distributed  test (distributed,.txt`  |
| 51              | `1_win-vs2019-cuda11.7-py3  test (default, 1, 2, windows.8xla.txt` |
| 72              | `1_linux-bionic-cuda11.6-py3.9-gcc7  test (multigpu, 1, 1, li.txt` |

</details>

<details><summary>pull workflow</summary>

| warnings raised | log file name                                                      |
| --------------- | ------------------------------------------------------------------ |
| 0               | `1_linux-bionic-cuda11.6-py3.10-gcc7-bazel-test  build-and-te.txt` |
| 0               | `1_linux-bionic-cuda11.6-py3.10-gcc7  test (default, 1, 4, li.txt` |
| 0               | `1_linux-focal-py3.7-clang10-onnx  test (default, 1, 2, linux.txt` |
| 0               | `1_linux-focal-py3.7-clang7-asan  test (default, 1, 5, linux..txt` |
| 0               | `1_linux-vulkan-bionic-py3.7-clang9  test (default, 1, 1, lin.txt` |
| 0               | `2_linux-focal-py3.7-clang10-onnx  test (default, 2, 2, linux.txt` |
| 0               | `2_linux-focal-py3.7-clang7-asan  test (default, 2, 5, linux..txt` |
| 0               | `30_build-and-test.txt`                                            |
| 0               | `32_build-and-test.txt`                                            |
| 0               | `34_build-and-test.txt`                                            |
| 0               | `3_win-vs2019-cpu-py3  test (functorch, 1, 1, windows.4xlarge.txt` |
| 0               | `41_test.txt`                                                      |
| 0               | `46_test.txt`                                                      |
| 0               | `49_test.txt`                                                      |
| 0               | `52_test.txt`                                                      |
| 0               | `55_test.txt`                                                      |
| 0               | `58_test.txt`                                                      |
| 0               | `5_linux-focal-py3.7-gcc7  test (functorch, 1, 1, linux.2xlar.txt` |
| 0               | `61_test.txt`                                                      |
| 0               | `64_test.txt`                                                      |
| 0               | `6_linux-focal-py3.7-clang7-asan  test (functorch, 1, 1, linu.txt` |
| 0               | `7_linux-bionic-py3.7-clang9  test (functorch, 1, 1, linux.2x.txt` |
| 0               | `8_linux-bionic-cuda11.6-py3.10-gcc7  test (functorch, 1, 1,.txt`  |
| 0               | `8_linux-focal-py3.7-gcc7  test (backwards_compat, 1, 1, linu.txt` |
| 1               | `3_linux-focal-py3.7-gcc7  test (distributed, 1, 2, linux.2xl.txt` |
| 1               | `9_linux-bionic-cuda11.6-py3.10-gcc7  test (deploy, 1, 1, lin.txt` |
| 2               | `7_linux-focal-py3.7-gcc7  test (jit_legacy, 1, 1, linux.2xla.txt` |
| 5               | `4_linux-focal-py3.7-gcc7  test (distributed, 2, 2, linux.2xl.txt` |
| 5               | `7_linux-bionic-cuda11.6-py3.10-gcc7  test (distributed, 3, 3.txt` |
| 10              | `5_linux-focal-py3.7-clang7-asan  test (default, 5, 5, linux..txt` |
| 13              | `3_linux-focal-py3.7-clang7-asan  test (default, 3, 5, linux..txt` |
| 15              | `1_linux-bionic-py3_7-clang8-xla  test (xla, 1, 1, linux.2xla.txt` |
| 19              | `6_linux-bionic-cuda11.6-py3.10-gcc7  test (distributed, 2, 3.txt` |
| 20              | `2_linux-bionic-cuda11.6-py3.10-gcc7  test (default, 2, 4, li.txt` |
| 20              | `2_linux-bionic-py3.7-clang9  test (default, 2, 2, linux.2xla.txt` |
| 21              | `1_win-vs2019-cpu-py3  test (default, 1, 2, windows.4xlarge).txt`  |
| 23              | `3_linux-bionic-cuda11.6-py3.10-gcc7  test (default, 3, 4, li.txt` |
| 24              | `4_linux-bionic-cuda11.6-py3.10-gcc7  test (default, 4, 4, li.txt` |
| 26              | `4_linux-bionic-py3.7-clang9  test (crossref, 2, 2, linux.2xl.txt` |
| 27              | `2_linux-focal-py3.7-gcc7  test (default, 2, 2, linux.2xlarge.txt` |
| 34              | `6_linux-focal-py3.7-gcc7  test (docs_test, 1, 1, linux.2xlar.txt` |
| 36              | `3_linux-bionic-py3.7-clang9  test (crossref, 1, 2, linux.2xl.txt` |
| 39              | `1_linux-focal-py3.7-gcc7  test (default, 1, 2, linux.2xlarge.txt` |
| 40              | `2_win-vs2019-cpu-py3  test (default, 2, 2, windows.4xlarge).txt`  |
| 42              | `4_linux-focal-py3.7-clang7-asan  test (default, 4, 5, linux..txt` |
| 46              | `1_linux-bionic-py3.7-clang9  test (default, 1, 2, linux.2xla.txt` |
| 54              | `5_linux-bionic-cuda11.6-py3.10-gcc7  test (distributed, 1, 3.txt` |
| 69              | `2_linux-docs  build-docs (python, linux.2xlarge, 30).txt`         |
| 119             | `6_linux-bionic-py3.7-clang9  test (dynamo, 2, 2, linux.2xlar.txt` |
| 216             | `5_linux-bionic-py3.7-clang9  test (dynamo, 1, 2, linux.2xlar.txt` |

</details>

<details><summary>trunk workflow</summary>

| warnings raised | log file name                                                      |
| --------------- | ------------------------------------------------------------------ |
| 0               | `13_linux-bionic-cuda11.7-py3.10-gcc7  test (distributed, 3,.txt`  |
| 0               | `1_android-emulator-build-test  build-and-test.txt`                |
| 0               | `1_linux-bionic-cuda11.7-py3.10-gcc7  test (default, 1, 4, li.txt` |
| 0               | `1_linux-bionic-py3.7-clang9-slow  test (slow, 1, 1, linux.2x.txt` |
| 0               | `1_linux-focal-py3.7-clang7-tsan  test (tsan, 1, 1, linux.2xl.txt` |
| 0               | `32_build-and-test.txt`                                            |
| 0               | `35_test.txt`                                                      |
| 0               | `38_test.txt`                                                      |
| 0               | `3_macos-12-py3-arm64  test (functorch, 1, 1, macos-m1-12).txt`    |
| 0               | `3_macos-12-py3-x86-64  test (functorch, 1, 1, macos-12).txt`      |
| 0               | `41_test.txt`                                                      |
| 0               | `44_test.txt`                                                      |
| 0               | `47_test.txt`                                                      |
| 0               | `50_test.txt`                                                      |
| 0               | `52_Run MPS tests.txt`                                             |
| 0               | `55_test.txt`                                                      |
| 0               | `58_test.txt`                                                      |
| 0               | `5_cuda11.6-py3.10-gcc7-sm86  test (functorch, 1, 1, linux.g5.txt` |
| 0               | `5_linux-bionic-cuda11.7-py3.10-gcc7  test (functorch, 1, 1,.txt`  |
| 0               | `60_test.txt`                                                      |
| 0               | `6_linux-bionic-cuda11.7-py3.10-gcc7  test (slow, 1, 2, linux.txt` |
| 0               | `6_win-vs2019-cuda11.6-py3  test (functorch, 1, 1, windows.8x.txt` |
| 0               | `7_linux-bionic-cuda11.7-py3.10-gcc7  test (slow, 2, 2, linux.txt` |
| 2               | `1_macos-12-py3-arm64-mps  Run MPS tests.txt`                      |
| 3               | `10_linux-bionic-cuda11.7-py3.10-gcc7  test (jit_legacy, 1, 1.txt` |
| 3               | `1_win-vs2019-cuda11.6-py3  test (default, 1, 5, windows.8xla.txt` |
| 6               | `3_win-vs2019-cuda11.6-py3  test (default, 3, 5, windows.8xla.txt` |
| 6               | `4_cuda11.6-py3.10-gcc7-sm86  test (default, 4, 4, linux.g5.4.txt` |
| 9               | `5_win-vs2019-cuda11.6-py3  test (default, 5, 5, windows.8xla.txt` |
| 12              | `4_linux-bionic-cuda11.7-py3.10-gcc7  test (default, 4, 4, li.txt` |
| 14              | `2_cuda11.6-py3.10-gcc7-sm86  test (default, 2, 4, linux.g5.4.txt` |
| 14              | `4_win-vs2019-cuda11.6-py3  test (default, 4, 5, windows.8xla.txt` |
| 17              | `2_linux-focal-rocm5.2-py3.8  test (default, 2, 2, linux.rocm.txt` |
| 18              | `3_cuda11.6-py3.10-gcc7-sm86  test (default, 3, 4, linux.g5.4.txt` |
| 20              | `12_linux-bionic-cuda11.7-py3.10-gcc7  test (distributed, 2,.txt`  |
| 20              | `2_macos-12-py3-arm64  test (default, 2, 2, macos-m1-12).txt`      |
| 24              | `1_parallelnative-linux-focal-py3.7-gcc7  test (default, 1, 2.txt` |
| 25              | `3_linux-bionic-cuda11.7-py3.10-gcc7  test (default, 3, 4, li.txt` |
| 27              | `7_win-vs2019-cuda11.6-py3  test (force_on_cpu, 1, 1, windows.txt` |
| 29              | `1_cuda11.6-py3.10-gcc7-sm86  test (default, 1, 4, linux.g5.4.txt` |
| 30              | `2_linux-bionic-cuda11.7-py3.10-gcc7  test (default, 2, 4, li.txt` |
| 31              | `8_linux-bionic-cuda11.7-py3.10-gcc7  test (nogpu_AVX512, 1,.txt`  |
| 31              | `9_linux-bionic-cuda11.7-py3.10-gcc7  test (nogpu_NO_AVX2, 1,.txt` |
| 33              | `2_win-vs2019-cuda11.6-py3  test (default, 2, 5, windows.8xla.txt` |
| 42              | `2_parallelnative-linux-focal-py3.7-gcc7  test (default, 2, 2.txt` |
| 44              | `1_macos-12-py3-arm64  test (default, 1, 2, macos-m1-12).txt`      |
| 46              | `1_linux-focal-rocm5.2-py3.8  test (default, 1, 2, linux.rocm.txt` |
| 46              | `2_macos-12-py3-x86-64  test (default, 2, 2, macos-12).txt`        |
| 58              | `11_linux-bionic-cuda11.7-py3.10-gcc7  test (distributed, 1,.txt`  |
| 75              | `1_macos-12-py3-x86-64  test (default, 1, 2, macos-12).txt`        |

</details>


Most of these warnings are raised because test code explicitly uses the deprecated functions, and there are multiple warnings in one job because they get triggered by different test files, which are run as separate processes. For instance, `test_torch.py` may generate one, and then `test_autograd.py` may generate another one. I don't consider such instances to be actual duplicate warnings since they come from different processes, and there is not a large number of them anyway.

But some of the logs show actual duplicates--multiple warnings generated by the execution of just one test file. There seem to be only a few root causes remaining for these. Here are some that I looked into, but haven't looked into all of them yet.

<details><summary>click to expand</summary>

75 from `1_macos-12-py3-x86-64  test (default, 1, 2, macos-12).txt`
  * 6 from `test_serialization.py`
  * ~50 from `test_quantization.py`

58 from `11_linux-bionic-cuda11.7-py3.10-gcc7  test (distributed, 1,.txt`
  * 5 from `distributed/test_dynamo_distributed.py`
  * 52 from `distributed/fsdp/test_fsdp_mixed_precision.py`

46 from `2_macos-12-py3-x86-64  test (default, 2, 2, macos-12).txt`
  * 9 from `test_view_ops.py -k test_as_strided_gradients_cpu` because of `warnings.catch_warnings()`
  * 9 from `test_view_ops.py -k test_as_strided_gradients_lazy` because of `warnings.catch_warnings()`
  * 6 from `test_torch.py`

46 from `1_linux-focal-rocm5.2-py3.8  test (default, 1, 2, linux.rocm.txt`
  * 6 from `test_serialization.py`
  * 9 from `test_view_ops.py -k test_as_strided_gradients_cuda` because of `warnings.catch_warnings()`
  * 8 from `test_nn.py -k test_share_memory`

44 from `1_macos-12-py3-arm64  test (default, 1, 2, macos-m1-12).txt`
  * 9 from `test_view_ops.py -k test_as_strided_gradients_cpu` because of `warnings.catch_warnings()`
  * 6 from `test_torch.py`


42 `2_parallelnative-linux-focal-py3.7-gcc7  test (default, 2, 2.txt`
  * 6 from `test_serialization.py`
  * 9 from `test_view_ops.py -k test_as_strided_gradients_cpu` because of `warnings.catch_warnings()`
  * 9 from `test_view_ops.py -k test_as_strided_gradients_lazy` because of `warnings.catch_warnings()`

33 `2_win-vs2019-cuda11.6-py3  test (default, 2, 5, windows.8xla.txt`
  * 5 from `test_serialization.py`
  * 9 from `test_view_ops.py -k test_as_strided_gradients_cuda` because of `warnings.catch_warnings()`
  * 8 from `test_nn.py -k test_share_memory`

31 `9_linux-bionic-cuda11.7-py3.10-gcc7  test (nogpu_NO_AVX2, 1,.txt`
  * 5 from `test_serialization.py`
  
</details>

ezyang commented:
re the crossref failures, part of the crossref process involves monkeying around with storage api. Right now it's using the deprecated api which we should fix

```
            self.storage_memo[swr] = (
                callback(
                    lambda: torch.empty(s.size(), dtype=torch.uint8, device="meta")
                )
                .storage()
                .untyped()
```

Probably dynamo is similar.

kurtamohler commented:
@ezyang, where is that code snippet from? Having trouble finding it

ezyang commented:
torch/_subclasses/meta_utils.py

kurtamohler commented:
Oh, I already changed that to use the internal functions, so that's not where the duplicates are coming from.

And actually, I misspoke--the latest update I made fixed most of the duplicates from crossref. As the table for the pull workflow above shows, the two crossref jobs only have 26 and 36 warnings now--and looking at the logs, I can see that these are mostly due to `warning.catch_warnings()`.

The two dynamo jobs, on the other hand, still have the highest warning counts. Here's a summary of most of the duplicates from one of them:

216 from `5_linux-bionic-py3.7-clang9  test (dynamo, 1, 2, linux.2xlar.txt`
  * 4 from `PYTORCH_TEST_WITH_DYNAMO=1 python test/test_indexing.py -k test_advancedindex_cpu_float64`
  * 82 from `PYTORCH_TEST_WITH_DYNAMO=1 python test/test_view_ops.py`
    - 12 from `test_as_strided_gradients_cpu` because of `warnings.catch_warnings()`
    - 5 from `test_flatten_view_cpu`
    - 10 from `test_as_strided_gradients_lazy` because of `warnings.catch_warnings()`
  * 13 from `PYTORCH_TEST_WITH_DYNAMO=1 test_bundled_inputs.py`

Looking at tracebacks added to the warning messages, most (if not all) of these are triggered by calls to the public TypedStorage function calls in the test code, not by any public function calls in `torch/_dynamo/`. I don't know why enabling dynamo makes this happen--it seems possible that something is resetting the warning filter

kurtamohler commented:
The `warnings.catch_warnings()` issue doesn't seem to have a workaround at the moment--none has been suggested on the issue (https://github.com/python/cpython/issues/73858), which was opened in 2018

kurtamohler commented:
There is somewhat of a workaround, but it won't work as shown below in practice because if there are warnings raised inside the `catch_warnings` context, then the version of the restored registry has to be incremented accordingly. I haven't found a way to determine what version number the `warning` module expects, but there might be some way

```python
import warnings

for _ in range(3):
    try:
        saved_registry = __warningregistry__
    except NameError:
        saved_registry = None

    with warnings.catch_warnings():
        pass

    if saved_registry is not None:
        saved_registry['version'] += 2
        __warningregistry__ = saved_registry

    warnings.warn('warning')
    warnings.warn(RuntimeWarning('another warning'))
```

ezyang commented:
dummy warning after catch warning doesn't seem like a huge problem

kurtamohler commented:
Yeah we can just filter out the dummy warning and it will still update the version number. This seems to basically work:

<details><summary>click to expand</summary>

```python
import warnings

class my_catch_warnings(warnings.catch_warnings):

    def __enter__(self):
        global __warningregistry__
        try:
            self._saved_registry = __warningregistry__.copy()
        except NameError:
            self._saved_registry = None

        super().__enter__()

    def __exit__(self, *args, **kwargs):
        super().__exit__(*args, **kwargs)

        if self._saved_registry is not None:
            global __warningregistry__
            dummy_message = 'dummy warning to find out warning registry version number'
            warnings.filterwarnings('ignore', message=dummy_message)
            warnings.warn(dummy_message)
            self._saved_registry['version'] = __warningregistry__['version']
            __warningregistry__ = self._saved_registry

for i in range(3):
    with my_catch_warnings():
        warnings.warn('inside catch_warnings')

    warnings.warn('this should only emit once')
    warnings.warn(RuntimeWarning('and this should only emit once too'))
```

Output:

```
/home/endoplasm/develop/python_stuff/catch_warnings_workaround.py:27: UserWarning: inside catch_warnings
  warnings.warn('inside catch_warnings')
/home/endoplasm/develop/python_stuff/catch_warnings_workaround.py:29: UserWarning: this should only emit once
  warnings.warn('this should only emit once')
/home/endoplasm/develop/python_stuff/catch_warnings_workaround.py:30: RuntimeWarning: and this should only emit once too
  warnings.warn(RuntimeWarning('and this should only emit once too'))
/home/endoplasm/develop/python_stuff/catch_warnings_workaround.py:27: UserWarning: inside catch_warnings
  warnings.warn('inside catch_warnings')
/home/endoplasm/develop/python_stuff/catch_warnings_workaround.py:27: UserWarning: inside catch_warnings
  warnings.warn('inside catch_warnings')
```

</details>

kurtamohler commented:
@pytorchbot merge -f "upstream CI failure"

kurtamohler commented:
BTW, before merging, I checked to make sure that the most recent update had similar warning counts in CI
